### PR TITLE
212: Connect adminstration to graph ql

### LINKED ISSUE #212: Connect Adminstration Backend to GraphQl endpoints
The adminstration backend should be connected to the graphql endpoints for creating new ehrenamtskarten.

Relates to #83.

### PR DESCRIPTION
**Pull request by Schedulaar**
Closes #212

### CODE REVIEW

File: administration/src/App.tsx [hunk 1]
```diff
@@ -4,36 +4,39 @@ import "normalize.css";
...
 import {HashRouter, Route} from "react-router-dom";
-import EakGeneration from "./components/generation/EakGeneration";
-import {createBinaryData} from "./util/protobuf";
-import {CardActivateModel} from "./generated/compiled";
+import GenerationController from "./components/generation/GenerationController";
+import styled from "styled-components";
 
 const client = new ApolloClient({
-    uri: 'https://api.ehrenamtskarte.app',
+    uri: 'http://api.ehrenamtskarte.app/verification',
```
Comment (maxammann) [inline administration/src/App.tsx:12]:
```suggestion
    uri: 'https://api.ehrenamtskarte.app/verification',
```

File: administration/src/App.tsx [hunk 2]
```diff
@@ -4,36 +4,39 @@ import "normalize.css";
...
 
+const Main = styled.div`
+  flex-grow: 1;
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+`
+
 
 function App() {
```
Comment (maxammann) [inline administration/src/App.tsx:24]: Use array syntax

File: administration/src/components/AppToaster.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { Position, Toaster } from "@blueprintjs/core";
+
+/** Singleton toaster instance. Create separate instances for different options. */
+export const AppToaster = Toaster.create({
```
Comment (maxammann) [inline administration/src/components/AppToaster.ts:4]:
```suggestion
export const AppToaster = Toaster.toast({
```

File: administration/src/components/generation/CardCreationModel.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+export interface CardCreationModel {
+    id: number,
+    forename: string,
+    surname: string,
+    expirationDate: Date,
+    cardType: CardType
+}
+
+export function isForenameValid(value: string) {
```
Comment (maxammann) [inline administration/src/components/generation/CardCreationModel.ts:13]: Use: => :raised_hands:

File: administration/src/components/generation/CardCreationModel.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    return value.length > 0 && value.length < MAX_NAME_LENGTH
+}
+
+export function isSurnameValid(value: string) {
```
Comment (maxammann) [inline administration/src/components/generation/CardCreationModel.ts:17]: Use: => :raised_hands:

File: administration/src/components/generation/CardCreationModel.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    return value.length > 0 && value.length < MAX_NAME_LENGTH
+}
+
+export function isExpirationDateValid(value: Date) {
```
Comment (maxammann) [inline administration/src/components/generation/CardCreationModel.ts:21]: Use: => :raised_hands:

File: administration/src/components/generation/GenerationController.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        return <GenerationForm cardCreationModels={cardCreationModels} setCardCreationModels={setCardCreationModels}
+                               confirm={confirm}/>
+    else if (mode === Mode.loading)
+        return <Spinner/>
+    else // (mode === Mode.finished)
+        return <GenerationFinished reset={() => {
+            setCardCreationModels([]);
+            setMode(Mode.input);
+        }}/>
+};
```
Comment (maxammann) [inline administration/src/components/generation/GenerationController.tsx:56]:
```suggestion
    if (mode === Mode.input)
        return <GenerationForm cardCreationModels={cardCreationModels} setCardCreationModels={setCardCreationModels}
                               confirm={confirm}/>
    else if (mode === Mode.loading)
        return <Spinner/>
    else // (mode === Mode.finished)
        return <GenerationFinished reset={() => {
            setCardCreationModels([]);
            setMode(Mode.input);
        }}/>
};

```

Use the Curly force! (missing {})

File: administration/src/components/generation/GenerationForm.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+const GenerationForm = (props: Props) => {
+    const {cardCreationModels, setCardCreationModels} = props
+    const addForm = () => setCardCreationModels([...cardCreationModels, createEmptyCard()]);
+    const updateModel = (oldModel: CardCreationModel, newModel: CardCreationModel | null) => {
+        if (newModel === null)
+            props.setCardCreationModels(cardCreationModels.filter(model => model !== oldModel));
+        else
+            setCardCreationModels(cardCreationModels.map(model => model === oldModel ? newModel : model))
```
Comment (maxammann) [inline administration/src/components/generation/GenerationForm.tsx:68]: Some curly-love missing here :heart:

File: administration/src/components/generation/GenerationForm.tsx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    setCardCreationModels: (models: CardCreationModel[]) => void,
+    confirm: () => void
+}
+
+const GenerationForm = (props: Props) => {
+    const {cardCreationModels, setCardCreationModels} = props
+    const addForm = () => setCardCreationModels([...cardCreationModels, createEmptyCard()]);
+    const updateModel = (oldModel: CardCreationModel, newModel: CardCreationModel | null) => {
+        if (newModel === null)
+            props.setCardCreationModels(cardCreationModels.filter(model => model !== oldModel));
```
Comment (maxammann) [inline administration/src/components/generation/GenerationForm.tsx:66]:
```suggestion
            setCardCreationModels(cardCreationModels.filter(model => model !== oldModel));
```

File: administration/src/components/generation/GenerationForm.tsx [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        else
+            setCardCreationModels(cardCreationModels.map(model => model === oldModel ? newModel : model))
+    }
+
+    const allCardsValid = cardCreationModels.reduce((acc, model) => acc && isValid(model), true)
+
+    return (
+        <>
+            <Prompt message={"Die Eingaben werden verworfen, falls Sie fortfahren."}
```
Comment (maxammann) [inline administration/src/components/generation/GenerationForm.tsx:75]:
```suggestion
            <Prompt message={"Falls Sie fortfahren, werden alle Eingaben verworfen."}
```

File: administration/src/components/generation/downloadDataUri.tsx [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+function downloadDataUri(blob: Blob, fileName: string) {
+    const newBlob = new Blob([blob], {type: "application/pdf"})
+    if (!isIE11()) {
+        const data = window.URL.createObjectURL(newBlob);
+        const downloadLink = document.createElement("a");
+        downloadLink.href = data;
+        downloadLink.download = fileName;
+        downloadLink.click();
+    } else {
+        window.navigator.msSaveBlob(newBlob, fileName)
```
Comment (maxammann) [inline administration/src/components/generation/downloadDataUri.tsx:17]: Microsoft boss to employee: `prefix that function with ms or you are fired!"

File: administration/src/index.tsx [hunk 12]
```diff
@@ -1,14 +1,13 @@
...
 import './index.css';
 import App from './App';
 import reportWebVitals from './reportWebVitals';
 
 ReactDOM.render(
-  <React.StrictMode>
-    <App />
-  </React.StrictMode>,
-  document.getElementById('root')
+    React.createElement(App), // Don't use <App />, otherwise Babel will destroy IE11 functionality
```
Comment (maxammann) [inline administration/src/index.tsx:9]: Babel will fuck you up!

File: administration/src/util/generateHashFromHashModel.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+export interface HashModel {
+    fullName: string,
+    region: number,
+    randomBytes: Uint8Array,
+    cardType: CardActivateModel.CardType,
+}
+
+const generateHashFromHashModel = async (hashModel: HashModel) => {
+    // todo: replace this routine with e.g. a protobuf hashModel for reliable hashes
```
Comment (maxammann) [inline administration/src/util/generateHashFromHashModel.ts:12]:
```suggestion
    // todo: replace this routine with a custom encoding for reliable hashes
```

**Review by maxammann — APPROVED** [review]

File: administration/src/util/generateHashFromHashModel.ts (hunk 13 above)
Comment (Schedulaar) [reply administration/src/util/generateHashFromHashModel.ts:12]: Update comment

**Review by maxammann — APPROVED** [review]

**Review by maxammann — APPROVED** [review]

File: administration/apollo.config.js [hunk 14]
```diff
@@ -1,6 +1,6 @@
 module.exports = {
   client: {
     name: "Ehrenamtskarte",
-    service: { name: "ehrenamtskarte-graphql-service", localSchemaFile: "../specs/backend-api.graphql" }
+    service: { name: "ehrenamtskarte-graphql-service", localSchemaFile: "../specs/verification-api.graphql" }
```
Comment (elKei24) [inline administration/apollo.config.js:4]: We should remember to undo this if we merge #215 before this one

File: administration/src/App.tsx [hunk 15]
```diff
@@ -4,39 +4,39 @@ import "normalize.css";
...
 import {HashRouter, Route} from "react-router-dom";
-import EakGeneration from "./components/generation/EakGeneration";
-import {createBinaryData} from "./util/protobuf";
-import {CardActivateModel} from "./generated/compiled";
+import GenerationController from "./components/generation/GenerationController";
+import styled from "styled-components";
 
 const client = new ApolloClient({
-    uri: 'https://api.ehrenamtskarte.app',
+    uri: 'https://api.ehrenamtskarte.app/verification',
```
Comment (elKei24) [inline administration/src/App.tsx:12]: Same here 😀

File: administration/src/components/generation/AddEakButton.tsx [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  background: white;
+  
+  :hover {
+    background: #f0f0f0;
+  }
+`
+
+const AddEakButton = (props: Props) => (
+    <StyledButton icon={<Icon style={{margin: 10}} icon={"add"} iconSize={20}/>} onClick={props.onClick}>
+        Karte Hinzufügen
```
Comment (elKei24) [inline administration/src/components/generation/AddEakButton.tsx:28]:
```suggestion
        Karte hinzufügen
```

File: administration/src/components/generation/CardCreationModel.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    id: number,
+    forename: string,
+    surname: string,
+    expirationDate: Date,
+    cardType: CardType
+}
+
+export const isForenameValid = (value: string) => value.length > 0 && value.length < MAX_NAME_LENGTH
+
+export const isSurnameValid = (value: string) => value.length > 0 && value.length < MAX_NAME_LENGTH
```
Comment (elKei24) [inline administration/src/components/generation/CardCreationModel.ts:15]: Just a trifle, but I would rather not have to identical functions for that as long as we do not handle forename and surname differently (which has a high chance of never happening)

File: administration/src/components/generation/EakForm.tsx [hunk 18]
```diff
@@ -32,44 +34,52 @@ const renderCardType: ItemRenderer<CardType> = (cardType, {handleClick, modifier
...
                 <FormGroup label="Ablaufdatum">
-                    <DateInput placeholder="Ablaufdatum" value={expirationDate}
+                    <DateInput placeholder="Ablaufdatum"
+                               value={props.model.expirationDate}
                                parseDate={str => new Date(str)}
-                               onChange={value => setExpirationDate(value)}
+                               onChange={value => props.onUpdate({...props.model, expirationDate: value})}
                                formatDate={date => date.toLocaleDateString()}
+                               maxDate={add(Date.now(), {years: 99})}
+                               minDate={new Date()}
```
Comment (elKei24) [inline administration/src/components/generation/EakForm.tsx:72]: I personally like the `Date.now()` better due to readability. Not worth putting effort in changing that however 😁

File: administration/src/components/generation/GenerationController.tsx [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    const [mode, setMode] = useState(Mode.input);
+    const confirm = async () => {
+        try {
+            setMode(Mode.loading)
+            const pdfDataUri = await generateCards(client, cardCreationModels)
+            downloadDataUri(pdfDataUri, "ehrenamtskarten.pdf")
+            setMode(Mode.finished)
+        } catch (e) {
+            console.error(e)
+            AppToaster.show({message: "Something went wrong.", intent: "danger"})
```
Comment (elKei24) [inline administration/src/components/generation/GenerationController.tsx:41]: I think the rest of our UI is in German, right?

```suggestion
            AppToaster.show({message: "Etwas ist schiefgegangen.", intent: "danger"})
```

File: administration/src/util/generateHashFromCardDetails.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+export interface CardDetails {
+    fullName: string,
+    regionId: number,
+    expirationDate: Long,
+    cardType: CardActivateModel.CardType,
+}
+
+const generateHashFromCardDetails = async (hashSecret: Uint8Array, cardDetails: CardDetails) => {
+    // todo: replace this routine with a custom encoding for reliable hashes
```
Comment (elKei24) [inline administration/src/util/generateHashFromCardDetails.ts:12]: Just to be sure: TODO is out of scope for this PR on purpose?

**Review by elKei24 — APPROVED** [review]
Did not go through it in too much detail, but looks good – the code, did not run it 😄

File: administration/src/util/generateHashFromCardDetails.ts (hunk 20 above)
Comment (elKei24) [reply administration/src/util/generateHashFromCardDetails.ts:12]: Ah, indeed, #219 handles this.

File: administration/src/components/generation/EakForm.tsx (hunk 18 above)
Comment (Schedulaar) [reply administration/src/components/generation/EakForm.tsx:72]: Yeah true. Problem is JS: Date.now() returns a number (I guess unix timestamp) and new Date() returns a Date object :sweat_smile:

Comment (elKei24) [reply administration/src/components/generation/EakForm.tsx:72]: Ah, okay. I guess we can not do much about that then 👍🏽😁

**Pull request merged**